controllers/machine_controller_phases.go

```diff
@@ -78,6 +78,13 @@ func (r *MachineReconciler) reconcileExternal(ctx context.Context, m *clusterv1.
 	obj, err := external.Get(r.Client, ref, m.Namespace)
 	if err != nil {
 		if apierrors.IsNotFound(err) {
+			if m.Status.InfrastructureReady && m.Status.GetTypedPhase() == clusterv1.MachinePhaseRunning {
```

Comment:
We should not check the phase - that is for human consumption. Checking InfrastructureReady is sufficient.

If we place this logic here, it applies to all external references - both the bootstrap ref and the infrastructure ref. @vincepri do you this is appropriate?

Reply:
I added the machine phase to make sure this block of code is executed only once.

Reply:
If we are not currently short-circuiting machine reconciliation if the error reason/error message are set, we need to be doing that.

Reply:
@ncdc I am not sure if I understood correctly, this check was allowing to setting the error reason/message and logging only once, but it did not short-circuit reconciliation

Reply:
In what situation would it be executing this more than once?

Reply:
In each call to reconcileExternal in my current version of the code?

Reply:
Ok, once you move this to reconcileInfrastructure, I think we also need to add code toward the top of Reconcile that checks machine.status.errorReason/errorMessage, and if either is set, it stops. This would need to be before the call to patch.NewHelper. Do you agree @vincepri?

Reply:
Yes, the short-circuit behavior is definitely missing. Let's follow-up with a different PR marked with ⚠️ for that